#!/usr/bin/env python3
"""
å°é¸½æ•°æ®å¹¿æ’­æ¨¡å—
ä»ç³»ç»Ÿè·å–å®æ—¶æ•°æ®ï¼Œé€šè¿‡UDPå¹¿æ’­åˆ°7701ç«¯å£
"""

import fcntl
import json  # ä¿ç•™ç”¨äºè°ƒè¯•ï¼ˆå¯é€‰ï¼‰
import math
import socket
import struct
import time
import traceback
import zlib
import gzip  # ğŸ”§ æ·»åŠ gzipå‹ç¼©æ”¯æŒ
from typing import Dict, Any, Optional, List, Tuple
import numpy as np

try:
    import msgpack
    MSGPACK_AVAILABLE = True
except ImportError:
    MSGPACK_AVAILABLE = False
    print("Warning: msgpack not available, falling back to JSON. Install with: pip install msgpack")

import cereal.messaging as messaging
from openpilot.common.realtime import Ratekeeper
from openpilot.system.hardware import PC


class XiaogeDataBroadcaster:
    def __init__(self):
        self.broadcast_port = 7701
        self.broadcast_ip = None
        self.sequence = 0

        # åˆå§‹åŒ–UDP socket
        self.udp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.udp_socket.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)

        # è®¢é˜…æ¶ˆæ¯
        self.sm = messaging.SubMaster([
            'carState',
            'modelV2',
            'radarState',
            'selfdriveState',
            'controlsState',
            'longitudinalPlan',
            'lateralPlan',  # æ·»åŠ  lateralPlan - å½“ activeLaneLine å¯ç”¨æ—¶ï¼Œè·¯å¾„æ•°æ®æ¥è‡ªè¿™é‡Œ
            'carControl',   # æ·»åŠ  carControl - ç”¨äºè·å–æ›²ç‡ä¿¡æ¯
            'carrotMan',
            # ç§»é™¤ 'can' - ç›²åŒºæ•°æ®ç›´æ¥ä»carStateè·å–
        ])

        # è·å–å¹¿æ’­åœ°å€
        self.broadcast_ip = self.get_broadcast_address()
        if self.broadcast_ip == '255.255.255.255':
            print("Warning: Could not determine network interface, using fallback broadcast address")

    def get_broadcast_address(self):
        """è·å–å¹¿æ’­åœ°å€"""
        interfaces = [b'br0', b'eth0', b'enp0s3'] if PC else [b'wlan0', b'eth0']

        for iface in interfaces:
            try:
                with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
                    ip = fcntl.ioctl(
                        s.fileno(),
                        0x8919,  # SIOCGIFADDR
                        struct.pack('256s', iface)
                    )[20:24]
                    broadcast_ip = socket.inet_ntoa(ip)
                    ip_parts = broadcast_ip.split('.')
                    ip_parts[3] = '255'
                    return '.'.join(ip_parts)
            except (OSError, Exception):
                continue

        return '255.255.255.255'

    def _capnp_list_to_list(self, capnp_list, max_items=None):
        """å°†capnpåˆ—è¡¨è½¬æ¢ä¸ºPythonåˆ—è¡¨"""
        if capnp_list is None:
            return []
        try:
            result = [float(x) for x in capnp_list]
            if max_items is not None:
                return result[:max_items]
            return result
        except (TypeError, AttributeError):
            return []

    def _capnp_enum_to_int(self, enum_value):
        """å°†capnpæšä¸¾è½¬æ¢ä¸ºæ•´æ•°"""
        try:
            return int(enum_value)
        except (TypeError, ValueError):
            return 0

    def _sample_array(self, arr: List[float], step: int = 2) -> List[float]:
        """
        å¯¹æ•°ç»„è¿›è¡Œé‡‡æ ·ï¼Œå‡å°‘æ•°æ®é‡
        step=1: ä¸é‡‡æ ·ï¼ˆå…¨éƒ¨ä¿ç•™ï¼‰
        step=2: æ¯éš”ä¸€ä¸ªç‚¹å–ä¸€ä¸ªï¼ˆä¿ç•™50%ï¼‰
        step=3: æ¯éš”ä¸¤ä¸ªç‚¹å–ä¸€ä¸ªï¼ˆä¿ç•™33%ï¼‰
        """
        if step <= 1 or len(arr) <= 1:
            return arr
        return arr[::step]

    def _calculate_curvature_from_path(self, x: List[float], y: List[float], sample: int = 4) -> Tuple[float, int]:
        """
        åŸºäºè·¯å¾„åæ ‡è®¡ç®—æ›²ç‡ï¼ˆå‚è€ƒ carrot_man.py:163-181ï¼‰
        ä½¿ç”¨ä¸‰ç‚¹æ³•è®¡ç®—æ›²ç‡ï¼šp1, p2, p3

        Args:
            x: è·¯å¾„çš„ x åæ ‡æ•°ç»„ï¼ˆè·ç¦»ï¼‰
            y: è·¯å¾„çš„ y åæ ‡æ•°ç»„ï¼ˆæ¨ªå‘åç§»ï¼‰
            sample: é‡‡æ ·é—´éš”ï¼ˆç”¨äºé€‰æ‹©ä¸‰ä¸ªç‚¹ï¼‰

        Returns:
            (curvature, direction): æ›²ç‡å€¼å’Œæ–¹å‘ï¼ˆ1=å·¦è½¬ï¼Œ-1=å³è½¬ï¼Œ0=ç›´é“ï¼‰
        """
        if len(x) < sample * 2 + 1 or len(y) < sample * 2 + 1:
            return 0.0, 0

        # é€‰æ‹©ä¸‰ä¸ªç‚¹ï¼šèµ·ç‚¹ã€ä¸­é—´ç‚¹ã€ç»ˆç‚¹
        # ä½¿ç”¨è·¯å¾„çš„å‰åŠéƒ¨åˆ†è®¡ç®—æ›²ç‡ï¼ˆæ›´æ¥è¿‘è½¦è¾†å½“å‰ä½ç½®ï¼‰
        max_idx = min(len(x) - sample * 2, 10)  # æœ€å¤šä½¿ç”¨å‰10ä¸ªç‚¹
        if max_idx < 1:
            return 0.0, 0

        # é€‰æ‹©ä¸‰ä¸ªç‚¹
        idx1 = 0
        idx2 = min(sample, max_idx - 1)
        idx3 = min(sample * 2, max_idx - 1)

        p1 = (x[idx1], y[idx1])
        p2 = (x[idx2], y[idx2])
        p3 = (x[idx3], y[idx3])

        # è®¡ç®—å‘é‡
        v1 = (p2[0] - p1[0], p2[1] - p1[1])
        v2 = (p3[0] - p2[0], p3[1] - p2[1])

        # è®¡ç®—å‰ç§¯
        cross_product = v1[0] * v2[1] - v1[1] * v2[0]
        len_v1 = math.sqrt(v1[0] ** 2 + v1[1] ** 2)
        len_v2 = math.sqrt(v2[0] ** 2 + v2[1] ** 2)

        if len_v1 * len_v2 == 0:
            return 0.0, 0

        # è®¡ç®—æ›²ç‡
        curvature = cross_product / (len_v1 * len_v2 * len_v1)

        # ç¡®å®šæ–¹å‘
        direction = 1 if curvature > 0 else (-1 if curvature < 0 else 0)

        return float(curvature), direction

    def collect_car_state(self, carState) -> Dict[str, Any]:
        """æ”¶é›†æœ¬è½¦çŠ¶æ€æ•°æ® - ä¼˜åŒ–ç‰ˆï¼ˆç§»é™¤å†—ä½™å­—æ®µï¼‰"""
        return {
            'vEgo': float(carState.vEgo),  # å®é™…é€Ÿåº¦
            'aEgo': float(carState.aEgo),  # åŠ é€Ÿåº¦
            'steeringAngleDeg': float(carState.steeringAngleDeg),  # æ–¹å‘ç›˜è§’åº¦
            'leftBlinker': bool(carState.leftBlinker),  # è½¬å‘ç¯
            'rightBlinker': bool(carState.rightBlinker),
            'brakePressed': bool(carState.brakePressed),  # åˆ¹è½¦
            'leftLatDist': float(carState.leftLatDist),  # è½¦é“è·ç¦»
            'rightLatDist': float(carState.rightLatDist),
            'leftLaneLine': int(carState.leftLaneLine),  # è½¦é“çº¿ç±»å‹
            'rightLaneLine': int(carState.rightLaneLine),
            'standstill': bool(carState.standstill),  # é™æ­¢çŠ¶æ€
            'leftBlindspot': bool(carState.leftBlindspot) if hasattr(carState, 'leftBlindspot') else False,  # å·¦ç›²åŒº
            'rightBlindspot': bool(carState.rightBlindspot) if hasattr(carState, 'rightBlindspot') else False,  # å³ç›²åŒº
            # ç§»é™¤ vEgoCluster - è¶…è½¦å†³ç­–ä¸éœ€è¦ä»ªè¡¨ç›˜é€Ÿåº¦
            # ç§»é™¤ vCruise - ä½¿ç”¨ longitudinalPlan.cruiseTarget ä»£æ›¿
        }

    def collect_model_data(self, modelV2, lateralPlan=None, carControl=None, controlsState=None) -> Dict[str, Any]:
        """
        æ”¶é›†æ¨¡å‹æ•°æ® - ä¼˜åŒ–ç‰ˆï¼ˆç§»é™¤å†—ä½™æ•°æ®ï¼Œåªä¿ç•™è¶…è½¦å†³ç­–æ‰€éœ€ï¼‰

        Args:
            modelV2: ModelV2 æ¶ˆæ¯å¯¹è±¡
            lateralPlan: LateralPlan æ¶ˆæ¯å¯¹è±¡ï¼ˆå¯é€‰ï¼Œå½“ activeLaneLine å¯ç”¨æ—¶ä½¿ç”¨ï¼‰
            carControl: CarControl æ¶ˆæ¯å¯¹è±¡ï¼ˆå¯é€‰ï¼Œç”¨äºè·å–æ›²ç‡ï¼‰
            controlsState: ControlsState æ¶ˆæ¯å¯¹è±¡ï¼ˆå¯é€‰ï¼Œç”¨äºåˆ¤æ–­ activeLaneLineï¼‰
        """
        data = {}

        # å‰è½¦æ£€æµ‹ - ä¿ç•™å…³é”®ä¿¡æ¯ï¼ˆè·ç¦»ã€é€Ÿåº¦ã€åŠ é€Ÿåº¦ã€ç½®ä¿¡åº¦ï¼‰
        if len(modelV2.leadsV3) > 0:
            lead0 = modelV2.leadsV3[0]
            data['lead0'] = {
                'x': float(lead0.x[0]) if len(lead0.x) > 0 else 0.0,  # è·ç¦»
                'v': float(lead0.v[0]) if len(lead0.v) > 0 else 0.0,  # é€Ÿåº¦
                'a': float(lead0.a[0]) if len(lead0.a) > 0 else 0.0,  # åŠ é€Ÿåº¦ï¼ˆç”¨äºåˆ¤æ–­å‰è½¦æ˜¯å¦åœ¨å‡é€Ÿï¼‰
                'prob': float(lead0.prob),  # ç½®ä¿¡åº¦
                # ç§»é™¤ 'y' (æ¨ªå‘ä½ç½®) - è¶…è½¦å†³ç­–ä¸éœ€è¦
            }
        else:
            data['lead0'] = {'x': 0.0, 'v': 0.0, 'a': 0.0, 'prob': 0.0}

        # ç¬¬äºŒå‰è½¦ - æ·»åŠ é€Ÿåº¦ä¿¡æ¯ï¼ˆç”¨äºåˆ¤æ–­è¶…è½¦ç©ºé—´ï¼‰
        if len(modelV2.leadsV3) > 1:
            lead1 = modelV2.leadsV3[1]
            data['lead1'] = {
                'x': float(lead1.x[0]) if len(lead1.x) > 0 else 0.0,  # è·ç¦»
                'v': float(lead1.v[0]) if len(lead1.v) > 0 else 0.0,  # é€Ÿåº¦ï¼ˆåˆ¤æ–­è¶…è½¦ç©ºé—´ï¼‰
                'prob': float(lead1.prob),  # ç½®ä¿¡åº¦
            }
        else:
            data['lead1'] = {'x': 0.0, 'v': 0.0, 'prob': 0.0}

        # ğŸ¯ è½¦é“çº¿æ•°æ® - å‘é€å®Œæ•´çš„4æ¡è½¦é“çº¿åæ ‡å’Œç½®ä¿¡åº¦ï¼ˆé‡‡æ ·ä»¥å‡å°‘æ•°æ®é‡ï¼‰
        # è½¦é“çº¿ç´¢å¼•ï¼š0=æœ€å·¦ä¾§, 1=å·¦è½¦é“çº¿, 2=å³è½¦é“çº¿, 3=æœ€å³ä¾§
        lane_lines = modelV2.laneLines
        data['laneLines'] = []
        # ğŸ”§ é‡‡æ ·æ­¥é•¿ï¼šä»3å¢åŠ åˆ°4ï¼ˆå‡å°‘75%æ•°æ®é‡ï¼‰ï¼Œè¿›ä¸€æ­¥å‡å°æ•°æ®åŒ…å¤§å°
        sample_step = 4
        for i in range(min(4, len(lane_lines))):
            lane_line = lane_lines[i]
            x_list = self._capnp_list_to_list(lane_line.x)
            y_list = self._capnp_list_to_list(lane_line.y)
            z_list = self._capnp_list_to_list(lane_line.z)
            data['laneLines'].append({
                'x': self._sample_array(x_list, sample_step),  # è·ç¦»æ•°ç»„ï¼ˆç±³ï¼‰- é‡‡æ ·
                'y': self._sample_array(y_list, sample_step),  # æ¨ªå‘åç§»æ•°ç»„ï¼ˆç±³ï¼‰- é‡‡æ ·
                'z': self._sample_array(z_list, sample_step),  # é«˜åº¦æ•°ç»„ï¼ˆç±³ï¼‰- é‡‡æ ·
            })
        # å¦‚æœå°‘äº4æ¡ï¼Œç”¨ç©ºæ•°æ®å¡«å……
        while len(data['laneLines']) < 4:
            data['laneLines'].append({'x': [], 'y': [], 'z': []})

        # è½¦é“çº¿ç½®ä¿¡åº¦ - å‘é€4ä¸ªå€¼
        if len(modelV2.laneLineProbs) >= 4:
            data['laneLineProbs'] = [
                float(modelV2.laneLineProbs[0]),  # æœ€å·¦ä¾§è½¦é“çº¿ç½®ä¿¡åº¦
                float(modelV2.laneLineProbs[1]),  # å·¦è½¦é“çº¿ç½®ä¿¡åº¦
                float(modelV2.laneLineProbs[2]),  # å³è½¦é“çº¿ç½®ä¿¡åº¦
                float(modelV2.laneLineProbs[3]),  # æœ€å³ä¾§è½¦é“çº¿ç½®ä¿¡åº¦
            ]
        elif len(modelV2.laneLineProbs) >= 2:
            # å…¼å®¹æ—§ç‰ˆæœ¬ï¼šåªæœ‰2ä¸ªå€¼ï¼Œå¡«å……ä¸º4ä¸ª
            data['laneLineProbs'] = [
                0.0,  # æœ€å·¦ä¾§ï¼ˆæœªçŸ¥ï¼‰
                float(modelV2.laneLineProbs[0]),  # å·¦è½¦é“çº¿
                float(modelV2.laneLineProbs[1]),  # å³è½¦é“çº¿
                0.0,  # æœ€å³ä¾§ï¼ˆæœªçŸ¥ï¼‰
            ]
        else:
            data['laneLineProbs'] = [0.0, 0.0, 0.0, 0.0]

        # è½¦é“å®½åº¦ã€åˆ°è·¯è¾¹ç¼˜è·ç¦»å’Œå˜é“çŠ¶æ€ - ä¿ç•™ï¼ˆè¶…è½¦å†³ç­–éœ€è¦ï¼‰
        meta = modelV2.meta
        data['meta'] = {
            'laneWidthLeft': float(meta.laneWidthLeft),  # å·¦è½¦é“å®½åº¦
            'laneWidthRight': float(meta.laneWidthRight),  # å³è½¦é“å®½åº¦
            'distanceToRoadEdgeLeft': float(meta.distanceToRoadEdgeLeft) if hasattr(meta, 'distanceToRoadEdgeLeft') else 0.0,  # åˆ°å·¦è·¯è¾¹ç¼˜è·ç¦»
            'distanceToRoadEdgeRight': float(meta.distanceToRoadEdgeRight) if hasattr(meta, 'distanceToRoadEdgeRight') else 0.0,  # åˆ°å³è·¯è¾¹ç¼˜è·ç¦»
            'laneChangeState': self._capnp_enum_to_int(meta.laneChangeState),
            'laneChangeDirection': self._capnp_enum_to_int(meta.laneChangeDirection),
        }

        # ğŸ¯ æ›²ç‡ä¿¡æ¯ - ä¼˜å…ˆä½¿ç”¨ carControl.actuators.curvatureï¼Œå¦åˆ™åŸºäºè·¯å¾„è®¡ç®—
        curvature_value = 0.0
        curvature_direction = 0
        curvature_obtained = False  # æ ‡è®°æ˜¯å¦å·²æˆåŠŸè·å–æ›²ç‡

        # æ–¹æ³•1ï¼šä¼˜å…ˆä» carControl è·å–æ›²ç‡ï¼ˆæœ€å‡†ç¡®ï¼‰
        if carControl and hasattr(carControl, 'actuators'):
            try:
                actuators = carControl.actuators
                if hasattr(actuators, 'curvature'):
                    curvature_value = float(actuators.curvature)
                    curvature_direction = 1 if curvature_value > 0 else (-1 if curvature_value < 0 else 0)
                    curvature_obtained = True  # æ ‡è®°å·²è·å–ï¼ˆå³ä½¿å€¼ä¸º0ï¼Œä¹Ÿè¡¨ç¤ºæ˜¯æœ‰æ•ˆçš„ç›´é“æ•°æ®ï¼‰
            except (AttributeError, TypeError):
                pass

        # æ–¹æ³•2ï¼šå¦‚æœ carControl æ²¡æœ‰æ›²ç‡ï¼Œå°è¯•ä» orientationRate è·å–ï¼ˆä¿æŒå‘åå…¼å®¹ï¼‰
        if not curvature_obtained:
            if hasattr(modelV2, 'orientationRate') and len(modelV2.orientationRate.z) > 0:
                orientation_rate_z = self._capnp_list_to_list(modelV2.orientationRate.z)
                if orientation_rate_z:
                    # æ‰¾åˆ°æœ€å¤§æ–¹å‘å˜åŒ–ç‡ï¼ˆè¡¨ç¤ºæœ€å¤§æ›²ç‡ç‚¹ï¼‰
                    max_index = max(range(len(orientation_rate_z)), key=lambda i: abs(orientation_rate_z[i]))
                    max_orientation_rate = orientation_rate_z[max_index]
                    curvature_value = float(max_orientation_rate)
                    curvature_direction = 1 if max_orientation_rate > 0 else -1
                    curvature_obtained = True

        # æ–¹æ³•3ï¼šå¦‚æœå‰ä¸¤ç§æ–¹æ³•éƒ½å¤±è´¥ï¼ŒåŸºäºè·¯å¾„åæ ‡è®¡ç®—æ›²ç‡
        if not curvature_obtained:
            # è·å–è·¯å¾„æ•°æ®ï¼ˆä¼˜å…ˆä½¿ç”¨ lateralPlanï¼Œå¦åˆ™ä½¿ç”¨ modelV2ï¼‰
            position = None
            if lateralPlan and hasattr(lateralPlan, 'position'):
                position = lateralPlan.position
            elif hasattr(modelV2, 'position'):
                position = modelV2.position

            if position:
                x_list = self._capnp_list_to_list(position.x)
                y_list = self._capnp_list_to_list(position.y)
                if len(x_list) >= 3 and len(y_list) >= 3:
                    curvature_value, curvature_direction = self._calculate_curvature_from_path(x_list, y_list)
                    curvature_obtained = True

        data['curvature'] = {
            'maxOrientationRate': curvature_value,  # æ›²ç‡å€¼ï¼ˆå¯èƒ½æ˜¯ curvature æˆ– orientationRateï¼‰
            'direction': curvature_direction,  # æ–¹å‘ï¼š1=å·¦è½¬ï¼Œ-1=å³è½¬ï¼Œ0=ç›´é“
        }

        # ğŸ¯ è·¯ç¼˜çº¿æ•°æ® - å‘é€2æ¡è·¯ç¼˜çº¿çš„åæ ‡å’Œæ ‡å‡†å·®ï¼ˆé‡‡æ ·ä»¥å‡å°‘æ•°æ®é‡ï¼‰
        if hasattr(modelV2, 'roadEdges') and len(modelV2.roadEdges) >= 2:
            road_edges = modelV2.roadEdges
            road_edge_stds = modelV2.roadEdgeStds if hasattr(modelV2, 'roadEdgeStds') else []

            data['roadEdges'] = []
            sample_step = 4  # ğŸ”§ é‡‡æ ·æ­¥é•¿ï¼šä»3å¢åŠ åˆ°4ï¼Œè¿›ä¸€æ­¥å‡å°æ•°æ®åŒ…å¤§å°
            for i in range(min(2, len(road_edges))):
                road_edge = road_edges[i]
                x_list = self._capnp_list_to_list(road_edge.x)
                y_list = self._capnp_list_to_list(road_edge.y)
                z_list = self._capnp_list_to_list(road_edge.z)
                data['roadEdges'].append({
                    'x': self._sample_array(x_list, sample_step),  # é‡‡æ ·
                    'y': self._sample_array(y_list, sample_step),  # é‡‡æ ·
                    'z': self._sample_array(z_list, sample_step),  # é‡‡æ ·
                })
            # å¦‚æœå°‘äº2æ¡ï¼Œç”¨ç©ºæ•°æ®å¡«å……
            while len(data['roadEdges']) < 2:
                data['roadEdges'].append({'x': [], 'y': [], 'z': []})

            # è·¯ç¼˜çº¿æ ‡å‡†å·®
            if len(road_edge_stds) >= 2:
                data['roadEdgeStds'] = [
                    float(road_edge_stds[0]),
                    float(road_edge_stds[1]),
                ]
            else:
                data['roadEdgeStds'] = [0.0, 0.0]
        else:
            data['roadEdges'] = [
                {'x': [], 'y': [], 'z': []},
                {'x': [], 'y': [], 'z': []},
            ]
            data['roadEdgeStds'] = [0.0, 0.0]

        # ğŸ¯ è·¯å¾„å¼•å¯¼æ•°æ® - ä¼˜å…ˆä½¿ç”¨ lateralPlan.positionï¼ˆå½“ useLaneLines å¯ç”¨æ—¶ï¼‰
        # å‚è€ƒ carrot.cc:1646-1650 å’Œ lateral_planner.py:231-243
        position = None
        use_lane_lines = False

        # æ£€æŸ¥æ˜¯å¦å¯ç”¨äº†è½¦é“çº¿æ¨¡å¼ï¼ˆä¼˜å…ˆä½¿ç”¨ lateralPlan.useLaneLinesï¼‰
        if lateralPlan and hasattr(lateralPlan, 'useLaneLines'):
            use_lane_lines = bool(lateralPlan.useLaneLines)
        elif controlsState and hasattr(controlsState, 'activeLaneLine'):
            # å¤‡é€‰æ–¹æ¡ˆï¼šä½¿ç”¨ controlsState.activeLaneLineï¼ˆå‘åå…¼å®¹ï¼‰
            use_lane_lines = bool(controlsState.activeLaneLine)

        # æ ¹æ® useLaneLines çŠ¶æ€é€‰æ‹©æ•°æ®æº
        if use_lane_lines and lateralPlan and hasattr(lateralPlan, 'position'):
            # å½“ useLaneLines å¯ç”¨æ—¶ï¼Œä½¿ç”¨ lateralPlan.position
            position = lateralPlan.position
        elif hasattr(modelV2, 'position'):
            # å¦åˆ™ä½¿ç”¨ modelV2.position
            position = modelV2.position

        if position:
            x_list = self._capnp_list_to_list(position.x)
            y_list = self._capnp_list_to_list(position.y)
            z_list = self._capnp_list_to_list(position.z)
            # é‡‡æ ·ä»¥å‡å°‘æ•°æ®é‡
            sample_step = 2
            data['position'] = {
                'x': self._sample_array(x_list, sample_step),
                'y': self._sample_array(y_list, sample_step),
                'z': self._sample_array(z_list, sample_step),
            }
        else:
            data['position'] = {'x': [], 'y': [], 'z': []}

        return data

    def collect_radar_data(self, radarState) -> Dict[str, Any]:
        """æ”¶é›†é›·è¾¾æ•°æ®ï¼ˆçº¯è§†è§‰æ–¹æ¡ˆä¹Ÿä¼šç”Ÿæˆè¿™äº›æ•°æ®ï¼‰"""
        data = {}

        # leadOneä¿¡æ¯
        leadOne = radarState.leadOne
        data['leadOne'] = {
            'dRel': float(leadOne.dRel),
            'vRel': float(leadOne.vRel),
            'vLead': float(leadOne.vLead),
            'vLeadK': float(leadOne.vLeadK),
            'status': bool(leadOne.status),
        }

        # leadTwoä¿¡æ¯
        if hasattr(radarState, 'leadTwo'):
            leadTwo = radarState.leadTwo
            data['leadTwo'] = {
                'dRel': float(leadTwo.dRel),
                'status': bool(leadTwo.status),
            }
        else:
            data['leadTwo'] = {'dRel': 0.0, 'status': False}

        # ä¾§æ–¹è½¦è¾†ä¿¡æ¯
        if hasattr(radarState, 'leadLeft'):
            leadLeft = radarState.leadLeft
            data['leadLeft'] = {
                'dRel': float(leadLeft.dRel) if leadLeft.status else 0.0,
                'vRel': float(leadLeft.vRel) if leadLeft.status else 0.0,
                'status': bool(leadLeft.status),
            }
        else:
            data['leadLeft'] = {'dRel': 0.0, 'vRel': 0.0, 'status': False}

        if hasattr(radarState, 'leadRight'):
            leadRight = radarState.leadRight
            data['leadRight'] = {
                'dRel': float(leadRight.dRel) if leadRight.status else 0.0,
                'vRel': float(leadRight.vRel) if leadRight.status else 0.0,
                'status': bool(leadRight.status),
            }
        else:
            data['leadRight'] = {'dRel': 0.0, 'vRel': 0.0, 'status': False}

        return data

    def collect_system_state(self, selfdriveState, controlsState) -> Dict[str, Any]:
        """æ”¶é›†ç³»ç»ŸçŠ¶æ€"""
        return {
            'enabled': bool(selfdriveState.enabled) if selfdriveState else False,
            'active': bool(selfdriveState.active) if selfdriveState else False,
            'longControlState': self._capnp_enum_to_int(controlsState.longControlState) if controlsState else 0,
        }

    def collect_carrot_data(self, carrotMan) -> Dict[str, Any]:
        """æ”¶é›† carrot å¯¼èˆªå’Œé™é€Ÿæ•°æ®"""
        return {
            'nRoadLimitSpeed': int(carrotMan.nRoadLimitSpeed) if hasattr(carrotMan, 'nRoadLimitSpeed') else 0,
            'desiredSpeed': int(carrotMan.desiredSpeed) if hasattr(carrotMan, 'desiredSpeed') else 0,
            'xSpdLimit': int(carrotMan.xSpdLimit) if hasattr(carrotMan, 'xSpdLimit') else 0,
            'xSpdDist': int(carrotMan.xSpdDist) if hasattr(carrotMan, 'xSpdDist') else 0,
            'xSpdType': int(carrotMan.xSpdType) if hasattr(carrotMan, 'xSpdType') else 0,
            'roadcate': int(carrotMan.roadcate) if hasattr(carrotMan, 'roadcate') else 0,  # é“è·¯ç±»å‹ï¼ˆé«˜é€Ÿ/å¿«é€Ÿè·¯/åŸå¸‚é“è·¯ï¼‰
        }

    # ç§»é™¤ collect_blindspot_data() - ç›²åŒºæ•°æ®å·²ç›´æ¥ä»carStateè·å–

    def create_packet(self, data: Dict[str, Any]) -> bytes:
        """
        åˆ›å»ºæ•°æ®åŒ…ï¼ŒåŒ…å«åºåˆ—å·ã€æ—¶é—´æˆ³å’Œæ ¡éªŒ
        ä½¿ç”¨ msgpack åºåˆ—åŒ–ï¼ˆæ¯” JSON å‡å°‘çº¦30%å¤§å°ï¼Œé€Ÿåº¦æå‡2-3å€ï¼‰
        """
        packet_data = {
            'version': 1,
            'sequence': self.sequence,
            'timestamp': time.time() * 1000,  # ğŸ”§ è½¬æ¢ä¸ºæ¯«ç§’ï¼ˆåŒ¹é…å®‰å“ç«¯æœŸæœ›ï¼‰
            'data': data
        }

        # ä½¿ç”¨ msgpack æ›¿ä»£ JSONï¼ˆå‡å°‘çº¦30%å¤§å°ï¼Œé€Ÿåº¦æå‡2-3å€ï¼‰
        # use_bin_type=True ç¡®ä¿å­—èŠ‚æ•°æ®è¢«æ­£ç¡®ç¼–ç ï¼Œå¯¹è·¨è¯­è¨€é€šä¿¡å¾ˆé‡è¦
        if MSGPACK_AVAILABLE:
            try:
                packet_bytes = msgpack.packb(packet_data, use_bin_type=True)
            except Exception as e:
                print(f"Warning: msgpack serialization failed, falling back to JSON: {e}")
                # å›é€€åˆ° JSON
                json_str = json.dumps(packet_data)
                packet_bytes = json_str.encode('utf-8')
        else:
            # å›é€€åˆ° JSONï¼ˆå¦‚æœ msgpack ä¸å¯ç”¨ï¼‰
            json_str = json.dumps(packet_data)
            packet_bytes = json_str.encode('utf-8')

        # ğŸ”§ æ·»åŠ gzipå‹ç¼©ï¼ˆå¯ä»¥å†å‡å°‘50-70%å¤§å°ï¼Œç‰¹åˆ«æ˜¯å¯¹äºé‡å¤æ•°æ®ï¼‰
        # å‹ç¼©çº§åˆ«ï¼š6ï¼ˆå¹³è¡¡å‹ç¼©ç‡å’Œé€Ÿåº¦ï¼‰
        try:
            compressed_bytes = gzip.compress(packet_bytes, compresslevel=6)
            # å¦‚æœå‹ç¼©åæ›´å°ï¼Œä½¿ç”¨å‹ç¼©æ•°æ®ï¼›å¦åˆ™ä½¿ç”¨åŸå§‹æ•°æ®
            if len(compressed_bytes) < len(packet_bytes):
                packet_bytes = compressed_bytes
                is_compressed = True
            else:
                is_compressed = False
        except Exception as e:
            print(f"Warning: gzip compression failed, using uncompressed data: {e}")
            is_compressed = False

        # æ·»åŠ CRC32æ ¡éªŒï¼ˆåœ¨å‹ç¼©åè®¡ç®—ï¼Œç¡®ä¿æ•°æ®å®Œæ•´æ€§ï¼‰
        checksum = zlib.crc32(packet_bytes) & 0xffffffff

        # æ•°æ®åŒ…æ ¼å¼: [å‹ç¼©æ ‡å¿—(1å­—èŠ‚)][æ ¡éªŒå’Œ(4å­—èŠ‚)][æ•°æ®é•¿åº¦(4å­—èŠ‚)][æ•°æ®]
        # å‹ç¼©æ ‡å¿—ï¼š0=æœªå‹ç¼©ï¼Œ1=å·²å‹ç¼©ï¼ˆgzipï¼‰
        packet = struct.pack('!BII', 1 if is_compressed else 0, checksum, len(packet_bytes)) + packet_bytes

        # æ£€æŸ¥æ•°æ®åŒ…å¤§å°
        # æ³¨æ„ï¼šç»è¿‡é‡‡æ ·ä¼˜åŒ–ï¼ˆsample_step=4ï¼‰ã€msgpackåºåˆ—åŒ–å’Œgzipå‹ç¼©åï¼Œæ•°æ®åŒ…å¤§å°åº”è¯¥æ˜¾è‘—å‡å°
        # å¦‚æœä»ç„¶è¶…è¿‡ 1400 å­—èŠ‚ï¼Œå¯èƒ½éœ€è¦è¿›ä¸€æ­¥é‡‡æ ·ï¼ˆå¢åŠ  sample_stepï¼‰æˆ–å®ç°åˆ†åŒ…å‘é€
        if len(packet) > 1400:  # ç•™ä¸€äº›ä½™é‡ï¼Œé¿å…è¶…è¿‡MTU
            compression_info = f" (å‹ç¼©={is_compressed})" if is_compressed else ""
            print(f"Warning: Packet size {len(packet)} bytes may exceed MTU{compression_info} (consider increasing sample_step)")

        return packet

    def broadcast_data(self):
        """ä¸»å¾ªç¯ï¼šæ”¶é›†æ•°æ®å¹¶å¹¿æ’­"""
        rk = Ratekeeper(10, print_delay_threshold=None)  # ğŸ”§ é™ä½åˆ°10Hzï¼ˆæ¯100msä¸€ä¸ªæ•°æ®åŒ…ï¼‰ï¼Œå‡å°‘ç½‘ç»œæ‹¥å¡å’Œæ•°æ®åŒ…ä¸¢å¤±

        print(f"XiaogeDataBroadcaster started, broadcasting to {self.broadcast_ip}:{self.broadcast_port}")

        while True:
            try:
                # æ€§èƒ½ç›‘æ§
                start_time = time.perf_counter()

                # æ›´æ–°æ‰€æœ‰æ¶ˆæ¯
                self.sm.update(0)

                # æ”¶é›†æ•°æ®
                data = {}

                # æœ¬è½¦çŠ¶æ€ - å§‹ç»ˆæ”¶é›†
                if self.sm.alive['carState']:
                    car_state = self.collect_car_state(self.sm['carState'])
                    # æ•°æ®éªŒè¯
                    if car_state.get('vEgo', 0) < 0:
                        print("Warning: Invalid vEgo value detected")
                    data['carState'] = car_state

                # æ¨¡å‹æ•°æ® - ä¼ å…¥ lateralPlan å’Œ carControl ä»¥æ”¯æŒå®Œæ•´åŠŸèƒ½
                if self.sm.alive['modelV2']:
                    lateral_plan = self.sm['lateralPlan'] if self.sm.alive['lateralPlan'] else None
                    car_control = self.sm['carControl'] if self.sm.alive['carControl'] else None
                    controls_state = self.sm['controlsState'] if self.sm.alive['controlsState'] else None
                    data['modelV2'] = self.collect_model_data(
                        self.sm['modelV2'],
                        lateralPlan=lateral_plan,
                        carControl=car_control,
                        controlsState=controls_state
                    )

                # é›·è¾¾æ•°æ®ï¼ˆçº¯è§†è§‰æ–¹æ¡ˆä¹Ÿä¼šæœ‰ï¼‰
                if self.sm.alive['radarState']:
                    data['radarState'] = self.collect_radar_data(self.sm['radarState'])

                # ç³»ç»ŸçŠ¶æ€
                if self.sm.alive['selfdriveState'] and self.sm.alive['controlsState']:
                    data['systemState'] = self.collect_system_state(
                        self.sm['selfdriveState'],
                        self.sm['controlsState']
                    )

                # çºµå‘è§„åˆ’æ•°æ®
                if self.sm.alive['longitudinalPlan']:
                    lp = self.sm['longitudinalPlan']
                    data['longitudinalPlan'] = {
                        'xState': self._capnp_enum_to_int(lp.xState),
                        'trafficState': self._capnp_enum_to_int(lp.trafficState),
                        'cruiseTarget': float(lp.cruiseTarget),
                        'hasLead': bool(lp.hasLead),
                    }

                # carrot å¯¼èˆªå’Œé™é€Ÿæ•°æ®
                if self.sm.alive['carrotMan']:
                    data['carrotMan'] = self.collect_carrot_data(self.sm['carrotMan'])

                # ç›²åŒºæ•°æ®å·²åŒ…å«åœ¨carStateä¸­ï¼Œæ— éœ€å•ç‹¬æ”¶é›†

                # æ€§èƒ½ç›‘æ§
                processing_time = time.perf_counter() - start_time
                if processing_time > 0.05:  # è¶…è¿‡50ms
                    print(f"Warning: Slow processing detected: {processing_time*1000:.1f}ms")

                # å¦‚æœæœ‰æ•°æ®åˆ™å¹¿æ’­
                if data:
                    packet = self.create_packet(data)

                    try:
                        self.udp_socket.sendto(packet, (self.broadcast_ip, self.broadcast_port))
                        self.sequence += 1

                        # æ¯100å¸§æ‰“å°ä¸€æ¬¡æ—¥å¿—ï¼ˆåŒ…å«æ•°æ®ç»Ÿè®¡ä¿¡æ¯ï¼‰
                        if self.sequence % 100 == 0:
                            # è®¡ç®—æ•°æ®ç»Ÿè®¡ä¿¡æ¯
                            stats_info = []
                            if 'modelV2' in data:
                                model_data = data['modelV2']
                                if 'laneLines' in model_data and len(model_data['laneLines']) > 0:
                                    lane_line_points = len(model_data['laneLines'][0]['x'])
                                    stats_info.append(f"laneLines: {lane_line_points}pts/line")
                                if 'roadEdges' in model_data and len(model_data['roadEdges']) > 0:
                                    road_edge_points = len(model_data['roadEdges'][0]['x'])
                                    stats_info.append(f"roadEdges: {road_edge_points}pts/edge")
                                if 'position' in model_data:
                                    position_points = len(model_data['position']['x'])
                                    stats_info.append(f"position: {position_points}pts")

                            # è¾“å‡ºæ—¥å¿—ï¼ˆæ˜¾ç¤ºåºåˆ—åŒ–æ ¼å¼å’Œå‹ç¼©ä¿¡æ¯ï¼‰
                            format_type = "msgpack" if MSGPACK_AVAILABLE else "JSON"
                            # æ£€æŸ¥æ•°æ®åŒ…æ˜¯å¦å‹ç¼©ï¼ˆé€šè¿‡æ£€æŸ¥æ•°æ®åŒ…æ ¼å¼ä¸­çš„å‹ç¼©æ ‡å¿—ï¼‰
                            try:
                                is_compressed = packet[0] == 1  # ç¬¬ä¸€ä¸ªå­—èŠ‚æ˜¯å‹ç¼©æ ‡å¿—
                                compression_info = " (gzipå‹ç¼©)" if is_compressed else ""
                            except:
                                compression_info = ""
                            stats_str = f" ({', '.join(stats_info)})" if stats_info else ""
                            print(f"Broadcasted {self.sequence} packets ({format_type}{compression_info}), last size: {len(packet)} bytes{stats_str}")
                    except Exception as e:
                        print(f"Failed to broadcast packet: {e}")

                rk.keep_time()

            except Exception as e:
                print(f"XiaogeDataBroadcaster error: {e}")
                traceback.print_exc()
                time.sleep(1)


def main():
    broadcaster = XiaogeDataBroadcaster()
    broadcaster.broadcast_data()


if __name__ == "__main__":
    main()